### DISCUSSION TITLE
LazyCollection from iterator without converting to array underneath?

### DISCUSSION THREAD

**Discussion by crissi** [laravel/framework#41735] [Q&A]
I know we do not allow generators as direct input to lazyCollection. Is the same problem with iterators?

It seems that the whole iterator is converted to an array when you use **new LazyCollection**.

Could we not allow iterators as input without converting it to an array?

**Comment by erikgaal** [comment]
Hi @crissi,

You _can_ actually start a `LazyCollection` with a generator function, like so:

```php
LazyCollection::make(function () {
    $number = 1;
	while (true) {
    	yield $number++;
    }
});
```

**Reply by crissi** [reply]
i know. But I was thinking it could do it automatically when given an iterator.

So I am thinking this code in EnumeratesValues:
![image](https://user-images.githubusercontent.com/3076578/160826985-17f71f35-9cd9-483e-8cf4-50b2b5d7ebf1.png)

could be replaced with this:



      
         elseif ($items instanceof Traversable) {
             return function() use ($items) {
                  foreach ($items as $row) {
                       yield $row;
                   }
             });

        }

**Reply by erikgaal** [reply]
But that function needs to return an _array_. It's exactly when you **want** to convert from an iterator to an array.

**Reply by crissi** [reply]
ahh okay, could it add to the the lazyCollection constructor instead.

**Reply by Anticom** [reply]
IMHO @crissi has a point here.

All terminal operations on the `LazyCollection` such as `each`, `combine` or `zip` make use of `makeIterator` which already supports `IteratorAggregate` (but not `Iterator` for some reason?)

If we don't look at the `getArrayableItems` but the ctor directly we could add support for `IteratorAggregate`(and possibly also `Iterator` as such:
```php
    public function __construct($source = null)
    {
        if ($source instanceof Closure || $source instanceof IteratorAggregate /*|| $source instanceof Iterator*/ || $source instanceof self) {
            $this->source = $source;
        } elseif (is_null($source)) {
            $this->source = static::empty();
        } elseif ($source instanceof Generator) {
            throw new InvalidArgumentException(
                'Generators should not be passed directly to LazyCollection. Instead, pass a generator function.'
            );
        } else {
            $this->source = $this->getArrayableItems($source);
        }
    }
```

**NB:** With the advent of `match` expressions in PHP 8.0 and current stable version of Laravel framework requiring `"php": "^8.2"` anyways, this could be written a bit more readable IMHO:

```php
    public function __construct($source = null)
    {
        $this->source = match(true) {
            $source instanceof Closure,
            $source instanceof IteratorAggregate,
            // In case we want & can support Iterator. (May be subsituted with checking for Traversable, if both Iterator & IteratorAggregate can be supported.)
            // $source instanceof Iterator,
            $source instanceof self => $source,
            is_null($source) =>  static::empty(),
            $source instanceof Generator => throw new InvalidArgumentException('Generators should not be passed directly to LazyCollection. Instead, pass a generator function.'),
            default => $this->getArrayableItems($source),
        };
    }
```

---

If a maintainer stumbles across this, lkm whether you'd be open for a PR. I'd be more than happy to contribute!
